File: 06. Backtracking/02. N-Queens/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Function to check if a queen can be placed at board[row][col]
bool isSafe(vector<string> &board, int row, int col, int n) {
    // Check column
    for (int i = 0; i < row; i++) {
        if (board[i][col] == 'Q') return false;
    }

    // Check upper-left diagonal
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j] == 'Q') return false;
    }

    // Check upper-right diagonal
    for (int i = row, j = col; i >= 0 && j < n; i--, j++) {
        if (board[i][j] == 'Q') return false;
    }

    return true;
}

// Backtracking function to solve N-Queens
void solveNQueens(int row, int n, vector<string> &board, vector<vector<string>> &solutions) {
    if (row == n) {
        solutions.push_back(board);
        return;
    }

    for (int col = 0; col < n; col++) {
        if (isSafe(board, row, col, n)) {
            board[row][col] = 'Q'; // Place queen
            solveNQueens(row + 1, n, board, solutions); // Recur for next row
            board[row][col] = '.'; // Backtrack
        }
    }
}

// Function to return all N-Queens solutions
vector<vector<string>> nQueens(int n) {
    vector<vector<string>> solutions;
    vector<string> board(n, string(n, '.')); // Initialize empty board
    solveNQueens(0, n, board, solutions);
    return solutions;
}

// Utility function to print the solutions
void printSolutions(const vector<vector<string>> &solutions) {
    for (const auto &solution : solutions) {
        for (const string &row : solution) {
            cout << row << "\n";
        }
        cout << "\n";
    }
}

int main() {
    int n = 8; // Change N to test different board sizes
    vector<vector<string>> solutions = nQueens(n);

    cout << "Solutions for " << n << "-Queens:\n";
    printSolutions(solutions);

    return 0;
}

File: 06. Backtracking/02. N-Queens/main.java
import java.util.*;

public class main {

    // Check if a queen can be placed safely
    private static boolean isSafe(List<String> board, int row, int col, int n) {
        for (int i = 0; i < row; i++) {
            if (board.get(i).charAt(col) == 'Q') return false;
        }

        for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
            if (board.get(i).charAt(j) == 'Q') return false;
        }

        for (int i = row, j = col; i >= 0 && j < n; i--, j++) {
            if (board.get(i).charAt(j) == 'Q') return false;
        }

        return true;
    }

    // Backtracking function to solve N-Queens
    private static void solveNQueens(int row, int n, List<String> board, List<List<String>> solutions) {
        if (row == n) {
            solutions.add(new ArrayList<>(board));
            return;
        }

        for (int col = 0; col < n; col++) {
            if (isSafe(board, row, col, n)) {
                char[] rowArray = board.get(row).toCharArray();
                rowArray[col] = 'Q';
                board.set(row, new String(rowArray));

                solveNQueens(row + 1, n, board, solutions);

                rowArray[col] = '.';
                board.set(row, new String(rowArray));
            }
        }
    }

    public static List<List<String>> nQueens(int n) {
        List<List<String>> solutions = new ArrayList<>();
        List<String> board = new ArrayList<>(Collections.nCopies(n, ".".repeat(n)));
        solveNQueens(0, n, board, solutions);
        return solutions;
    }

    public static void printSolutions(List<List<String>> solutions) {
        for (List<String> solution : solutions) {
            for (String row : solution) {
                System.out.println(row);
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        int n = 8;
        List<List<String>> solutions = nQueens(n);
        System.out.println("Solutions for " + n + "-Queens:");
        printSolutions(solutions);
    }
}
